### PR TITLE
chore(docs): Add more godoc to important APIs

### PR DESCRIPTION
**Pull request by yevgenypats**
As part of the v2 release we want to improve documentation for developers and for our plugin-sdk. 

Right now we have a bit of mix-match of tutorials and docs https://www.cloudquery.io/docs/developers/sdk/provider/configuration

What we want to achieve:
- SDK documentation in godoc together with docs - https://pkg.go.dev/github.com/cloudquery/plugin-sdk
- Tutorials - https://www.cloudquery.io/docs/developers/sdk/overview

There is still much more documentation to be added with some more examples but this should be a good start.

---

Use the following steps to ensure your PR is ready to be reviewed

- [ ] Read the [contribution guidelines](../blob/main/CONTRIBUTING.md) 🧑‍🎓
- [ ] Run `go fmt` to format your code 🖊
- [ ] Lint your changes via `golangci-lint run` 🚨 (install golangci-lint [here](https://golangci-lint.run/usage/install/#local-installation))
- [ ] Update or add tests 🧪
- [ ] Ensure the status checks below are successful ✅

### CODE REVIEW

File: schema/doc.go [hunk 1]
```diff
@@ -0,0 +1,2 @@
+// Package doc defines the main types supported by tables in source plugins
```
Comment (erezrokah) [inline schema/doc.go:1]:
```suggestion
// Package schema defines types supported by tables in source plugins
```

Or change the package name

File: serve/doc.go [hunk 2]
```diff
@@ -1,3 +1,4 @@
+// Package serve defines the main APIs to serve source and destination plugins
```
Comment (erezrokah) [inline serve/doc.go:1]:
```suggestion
// Package serve defines APIs to serve (invoke) source and destination plugins
```

File: specs/doc.go [hunk 3]
```diff
@@ -0,0 +1,3 @@
+// Package specs specs for source and destination plugins including
```
Comment (erezrokah) [inline specs/doc.go:1]: I'm not sure I understand what message this comment is trying to convey

File: specs/source.go [hunk 4]
```diff
@@ -7,19 +7,30 @@ import (
 	"github.com/xeipuuv/gojsonschema"
 )
 
-// Source is the shared configuration for all source plugins
+// Source is the spec for a source plugin
```
Comment (erezrokah) [inline specs/source.go:10]:
```suggestion
// Source defines the specification for a source plugin
```

File: plugins/source.go [hunk 5]
```diff
@@ -91,27 +93,32 @@ func (p *SourcePlugin) validate() error {
...
 func (p *SourcePlugin) Tables() schema.Tables {
 	return p.tables
 }
 
+// ExampleConfig returns an example configuration for this source plugin
 func (p *SourcePlugin) ExampleConfig() string {
 	return p.exampleConfig
 }
 
+// Name return the name of this plugin
```
Comment (erezrokah) [inline plugins/source.go:106]:
```suggestion
// Returns a human friendly name for the plugin
```

File: plugins/source.go [hunk 6]
```diff
@@ -91,27 +93,32 @@ func (p *SourcePlugin) validate() error {
...
 func (p *SourcePlugin) Name() string {
 	return p.name
 }
 
+// Version returns the version of this plugin
```
Comment (erezrokah) [inline plugins/source.go:111]:
```suggestion
// Returns the semantic valid version of the plugin
```

File: plugins/source.go [hunk 7]
```diff
@@ -91,27 +93,32 @@ func (p *SourcePlugin) validate() error {
...
 	return p.version
 }
 
+// SetLogger sets the logger for this plugin which will be used in Sync and all other function calls.
 func (p *SourcePlugin) SetLogger(log zerolog.Logger) {
 	p.logger = log
 }
 
-// Sync data from source to the given channel
+// Sync is syncing data from the requested tables in spec to the given channel
```
Comment (erezrokah) [inline plugins/source.go:121]:
```suggestion
// Sync retrieves data from the requested tables in the plugin's specification and sends the results via the provider `res` channel
```

File: plugins/source.go [hunk 8]
```diff
@@ -91,27 +93,32 @@ func (p *SourcePlugin) validate() error {
 	return nil
 }
 
+// Tables returns all supported tables by this source plugin
 func (p *SourcePlugin) Tables() schema.Tables {
 	return p.tables
 }
 
+// ExampleConfig returns an example configuration for this source plugin
```
Comment (erezrokah) [inline plugins/source.go:101]:
```suggestion
// Returns example configuration specific for this source plugin
```

File: specs/source.go [hunk 9]
```diff
@@ -7,19 +7,30 @@ import (
...
 type Source struct {
-	Name    string `json:"name,omitempty"`
+	// Name of the source plugin to use
```
Comment (erezrokah) [inline specs/source.go:12]:
```suggestion
	// Human friendly name of the plugin
```

File: specs/source.go [hunk 10]
```diff
@@ -7,19 +7,30 @@ import (
...
+	Name string `json:"name,omitempty"`
+	// Version of the source plugin to use
```
Comment (erezrokah) [inline specs/source.go:14]:
```suggestion
	// Valid semantic version of the plugin
```

File: specs/source.go [hunk 11]
```diff
@@ -7,19 +7,30 @@ import (
...
 	Version string `json:"version,omitempty"`
-	// Path is the path in the registry
+	// Path is the canonical path to the source plugin in a given registry
+	// For example:
+	// in github the path will be: org/repo
```
Comment (erezrokah) [inline specs/source.go:18]:
```suggestion
	// For the GitHub registry the path will be: org/repo
```

File: specs/source.go [hunk 12]
```diff
@@ -7,19 +7,30 @@ import (
...
+	// in local the path will be the path to the binary: ./path/to/binary
```
Comment (erezrokah) [inline specs/source.go:19]:
```suggestion
	// For the local registry the path will be the path to the binary: ./path/to/binary
```

File: specs/source.go [hunk 13]
```diff
@@ -7,19 +7,30 @@ import (
...
+	// in grpc the path will be the address of the grpc server: host:port
```
Comment (erezrokah) [inline specs/source.go:20]:
```suggestion
	// For the gRPC registry the path will be the address of the gRPC server: host:port
```

File: specs/source.go [hunk 14]
```diff
@@ -7,19 +7,30 @@ import (
...
-	Tables        []string    `json:"tables,omitempty"`
-	SkipTables    []string    `json:"skip_tables,omitempty"`
-	Destinations  []string    `json:"destinations,omitempty"`
-	Spec          interface{} `json:"spec,omitempty"`
+	// Registry can be github,local,grpc.
+	Registry      Registry `json:"registry,omitempty"`
+	MaxGoRoutines uint64   `json:"max_goroutines,omitempty"`
+	// Tables to sync from the source plugin
+	Tables []string `json:"tables,omitempty"`
+	// SkipTables mentions tables to skip from the source plugin. Useful if glob is used in Tables
```
Comment (erezrokah) [inline specs/source.go:27]:
```suggestion
	// SkipTables defines tables to skip when syncing data. Useful if a glob pattern is used in Tables
```

File: specs/source.go [hunk 15]
```diff
@@ -7,19 +7,30 @@ import (
...
+	SkipTables []string `json:"skip_tables,omitempty"`
+	// Destinations names of the destinations to sync the data to
```
Comment (erezrokah) [inline specs/source.go:29]:
```suggestion
	// Destinations are the names of destination plugins to send sync data to
```

File: specs/source.go [hunk 16]
```diff
@@ -7,19 +7,30 @@ import (
...
+	Destinations []string `json:"destinations,omitempty"`
+	// Spec is the specific spec defined by an exact source plugin.
```
Comment (erezrokah) [inline specs/source.go:31]:
```suggestion
	// Spec defines plugin specific configuration
```

**Review by erezrokah — APPROVED** [review]
Thanks @yevgenypats, I'm not sure all my comments makes sense so I would consider all non blocking.

I do think we should avoid docs that describe a standard operation of a get function  (e.g. `Name returns the name`)  and try to add more information when we can.

**Comment by yevgenypats** [comment]
> Thanks @yevgenypats, I'm not sure all my comments makes sense so I would consider all non blocking.
> 
> I do think we should avoid docs that describe a standard operation of a get function (e.g. `Name returns the name`) and try to add more information when we can.

Thanks @erezrokah ! Yeah it was my first try def room for improvement, all comments makes sense!

**Review by erezrokah — APPROVED** [review]

**Comment by erezrokah** [comment]
> Thanks @erezrokah ! Yeah it was my first try def room for improvement, all comments makes sense!

🚀 Looks good to me, maybe have a native English speaker cc @bbernays or @hermanschaaf look at this too

**Pull request merged**